File: CPP_files/soe.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring>
using namespace std;

void SieveOfEratosthenes(int n, ofstream &outputFile) {
    // Create a boolean array "prime[0..n]" and initialize
    // all entries it as true. A value in prime[i] will
    // finally be false if i is Not a prime, else true.
    bool prime[n + 1];
    memset(prime, true, sizeof(prime));

    for (int p = 2; p * p <= n; p++) {
        // If prime[p] is not changed, then it is a prime
        if (prime[p] == true) {
            // Update all multiples of p
            for (int i = p * p; i <= n; i += p)
                prime[i] = false;
        }
    }

    // Write prime numbers to output file
    for (int p = 2; p <= n; p++) {
        if (prime[p]) {
            outputFile << p;
            if (p < n) outputFile << ",";
        }
    }
}

int main() {
    // Check if "input.txt" exists
    ifstream inputFileCheck("input.txt");
    bool fileExists = inputFileCheck.good();
    inputFileCheck.close();

    // Open input file for appending
    ofstream inputFile("input.txt", ios_base::app);
    if (!inputFile) {
        cerr << "Error: Unable to open input file." << endl;
        return 1;
    }

    // If file doesn't exist, write test cases
    if (!fileExists) {
        vector<int> testCases = {10, 20, 30}; // Example test cases
        for (int n : testCases)
            inputFile << n << endl;
    }

    inputFile.close(); // Close input file

    // Open output file
    ofstream outputFile("output.txt");
    if (!outputFile) {
        cerr << "Error: Unable to create output file." << endl;
        return 1;
    }

    // Read from input file and write output to output file
    ifstream inputFileRead("input.txt");
    if (!inputFileRead) {
        cerr << "Error: Unable to open input file." << endl;
        return 1;
    }

    int n;
    while (inputFileRead >> n) {
        SieveOfEratosthenes(n, outputFile);
        outputFile << endl; // Separate test cases by newline
    }

    inputFileRead.close(); // Close input file
    outputFile.close();    // Close output file

    cout << "Output has been written to 'output.txt'" << endl;

    return 0;
}

File: PY_files/soe.py
def sieve_of_eratosthenes(n):
    # Create a boolean list "prime[0..n]" and initialize
    # all entries it as True. A value in prime[i] will
    # finally be False if i is Not a prime, else True.
    primes = [True] * (n + 1)
    p = 2
    while p * p <= n:
        if primes[p]:
            # Update all multiples of p
            for i in range(p * p, n + 1, p):
                primes[i] = False
        p += 1

    # Collect prime numbers
    prime_numbers = [p for p in range(2, n + 1) if primes[p]]
    return prime_numbers

def main():
    # Check if "input.txt" exists
    try:
        with open("input.txt", "r") as f:
            input_exists = True
    except FileNotFoundError:
        input_exists = False

    # Create input file with default test cases if it doesn't exist
    if not input_exists:
        test_cases = [10, 20, 30]  # Example test cases
        with open("input.txt", "w") as f:
            for case in test_cases:
                f.write(str(case) + "\n")
        print("Input file 'input.txt' has been created with default test cases")

    # Compute prime numbers for each test case
    with open("input.txt", "r") as f:
        test_cases = [int(line.strip()) for line in f]
    
    prime_results = [sieve_of_eratosthenes(n) for n in test_cases]

    # Write output to file
    with open("output.txt", "w") as f:
        for primes in prime_results:
            f.write(','.join(map(str, primes)) + '\n')

    print("Output has been written to 'output.txt'")

if __name__ == "__main__":
    main()
